src/libsyntax/parse/parser.rs

```diff
@@ -5319,6 +5355,163 @@ impl<'a> Parser<'a> {
+        // this will be reflected in the count:
+        //
+        // *Upcoming tokens:* `T as Foo>::Output>;`
+        // *Unmatched count:* 4
+        // `parse_path_segment` calls deep:* 3
+        //
+        // The parser will continue until reaching the first `>` - this will decrement the
+        // unmatched angle bracket count and return to the parent invocation of this function
+        // having succeeded in parsing:
+        //
+        // *Upcoming tokens:* `::Output>;`
+        // *Unmatched count:* 3
+        // *`parse_path_segment` calls deep:* 2
+        //
+        // This will continue until the next `>` character which will also return successfully
+        // to the parent invocation of this function and decrement the count:
+        //
+        // *Upcoming tokens:* `;`
+        // *Unmatched count:* 2
+        // *`parse_path_segment` calls deep:* 1
+        //
+        // At this point, this function will expect to find another matching `>` character but
+        // won't be able to and will return an error. This will continue all the way up the
+        // call stack until the first invocation:
+        //
+        // *Upcoming tokens:* `;`
+        // *Unmatched count:* 2
+        // *`parse_path_segment` calls deep:* 0
+        //
+        // In doing this, we have managed to work out how many unmatched leading left angle
+        // brackets there are, but we cannot recover as the unmatched angle brackets have
+        // already been consumed. To remedy this, whenever `parse_generic_args` is invoked, we
+        // make a snapshot of the current parser state and invoke it on that and inspect
+        // the result:
+        //
+        // - If success (ie. when it found a matching `>` character) then the snapshot state
+        //   is kept (this is required to propagate the count upwards).
+        //
+        // - If error and in was in a recursive call, then the snapshot state is kept (this is
+        //   required to propagate the count upwards).
+        //
+        // - If error and this was the first invocation (before any recursion had taken place)
+        //   then we choose not to keep the snapshot state - that way we haven't actually
+        //   consumed any of the `<` characters, but can still inspect the count from the
+        //   snapshot to know how many `<` characters to remove. Using this information, we can
+        //   emit an error and consume the extra `<` characters before attempting to parse
+        //   the generic arguments again (this time hopefullt successfully as the unmatched `<`
+        //   characters are gone).
+        //
+        // In practice, the recursion of this function is indirect and there will be other
+        // locations that consume some `<` characters - as long as we update the count when
+        // this happens, it isn't an issue.
+        let mut snapshot = self.clone();
```

Comment:
Even though I am to blame for the introduction of this concept in the codebase, I am always queasy about it when I see it 😬

I'll add a `try` run, I want to make sure this doesn't cause a performance regression in the compiler.

This is the parser, so it should be ok memory-wise, but still: measure twice, cut once.

The alternative would be to perform a lookahead _every time_ to look for `unmatched_angle_bracket_count`, but that would probably be worse.

Reply:
I initially tried a handful of other approaches, using snapshots as a last ditch attempt. 

I tried to limit the negative impacts of this approach as much as possible:

- `parse_generic_args` will be invoked at most once more than it was previously - this extra invocation is only after removing leading angle brackets (if we detect there are some) so the parser can recover and continue. 
- More memory will be used due to the snapshots:
    * This isn't as bad as the simplified example in the large comment might suggest. As mentioned at the end of that comment, the recursion of `parse_path_segment` is indirect and along the way, `<` tokens are consumed as type parsing is attempted - this directly limits the number of recursions that can be performed in `parse_path_segment` and therefore the number of snapshots required.
    * I found that in the most complex (and unrealistic) examples with 5+ leading brackets there were usually only two or three snapshots. In normal code, there is always going to be one snapshot for the first bracket but otherwise it would be relatively rare for there to be more.
- Code complexity has increased.